proposed/flexible-http.md

```diff
@@ -0,0 +1,149 @@
+- Better control over TLS handshakes.
+    - TBD.
+
+## APIs
+
+The APIs are split into three groups:
+
+- Low-level APIs
+- Compositional APIs
+- High-level APIs
+
+### Definitions
+
+- `Span<byte>` indicates a contiguous blob of unvalidated `byte` and may be used as a placeholder for all its variants e.g. `ReadOnlyMemory<byte>`.
+
+- `string` indicates a contiguous blob of `char` and may be used as a placeholder for all its variants e.g. `Span<char>`.
+
+- For discussion sake (actual API TBD), we will consider that this API consists of an abstraction and concrete implementations:
+    - `HttpConnection`, the abstraction.
+    - `Http1Connection`, `Http2Connection`, `Http3Connection`, the low-level implementations.
+
+## Low-level API
+
+Low-level APIs implement HTTP without any additional features or opinions. They return HTTP data direct to the user without transformation, and are built to be forward-compatible with new HTTP features (e.g. HTTP/2 and HTTP/3 extension frames) in the future.
+
+Performance is of high priority:
+
+- Operate on `Span<byte>` rather than `string` and `Uri`.
+    - Parameters will not be validated (to enable efficient forwarding), but helpers will be exposed to allow user to perform validation.
+- Zero allocations (steady-state).
+
+Ease of use is an explicit non-goal:
+
+- We assume users will know more about HTTP than `HttpClient` currently requires of them.
+- It is okay if it takes more code to use than `HttpClient`.
+
+These APIs are constructed on top of `Stream`. This has some implications worth noting:
+
+- Connections are not established by these APIs: the user must establish the connection.
+- Connecting via a proxy, or over TLS is not done by this API: the user must pass in a `Stream` that already has these established.
+
+### Header handling
+
+Headers are read and written "streaming", without going through a collection.
+
+"Prepared" headers will enable dynamic table compression and pre-validation/serialization equivalent to existing internal "known header" concept of `SocketsHttpHandler`.
+
+Huffman compression will be opt-in on a per-header basis.
+
+### Content handling
+
+Content will be read and written directly to the `HttpConnection`'s request type, without using `Stream` or a `HttpContent`-equivalent, to avoid the associated allocations.
+
+Expect Continue will not be implemented here, but full informational responses will be returned to allow the caller to implement their own Expect Continue processing.
+
+## Compositional APIs
+
+Compositional APIs are built on top of low-level APIs to provide additional features, as well as some opinionated implemenations.
+
+Every higher-level feature that `SocketsHttpHandler` currently supports should have an implementation here. This will mean one or more `HttpConnection` that implements some features:
+
+- Automatic decompression.
+- Automatic redirects.
+- Connection pooling.
+- Version negotiation.
+    - ALPN, Upgrade header, ALT-SVC.
```

Comment:
Re Upgrade:

Does this specifically refer to HTTP1.1->HTTP2 Upgrade? Is this something we think we would support?

Or is this referring to other Upgrade scenarios too?

Reply:
It'd be nice to add that so I didn't leave it out, but mostly I'm thinking ->HTTP3 via ALT-SVC.